src/Params/WaveTable.h

```diff
@@ -0,0 +1,659 @@
+    {
+        tensor_size_t consumed = 0;
+        for(Tensor<N-1, T>& sub : *this)
+        {
+            consumed += sub.debug_set_data_using_deep_copy(new_data + consumed);
+        }
+        return consumed;
+    }
+
+    Shape<N> debug_get_shape() const {
+        if(base_type::size()) {
+            // we do not store the whole shape, but only our own size,
+            // so calculate it iterting
+            return m_data[0].debug_get_shape().prepend_dim(base_type::size());
+        }
+        else {
+            // we can not compute the shape from m_data[0], because
+            // m_data has size 0 - so set it all to 0
+            Shape<N> res;
+            for(tensor_size_t i = 0; i < N; ++i) res.dim[i] = 0;
+            return res;
+        }
+    }
+};
+
+//! Tensor class for dimension 1
+template <class T>
+class Tensor<1, T> : public TensorBase<1, T>
+{
+    using base_type = TensorBase<1, T>;
+    T* m_data;
+
+protected:
+    Tensor() : m_data(nullptr) {}
+    void init_shape(const Shape<1>& shape)
+    {
+        assert(!m_data);
+        base_type::set_size(shape.dim[0]);
+        if(base_type::size())
+            m_data = new T[base_type::size()];
+    }
+public:
+    Tensor(tensor_size_t size) : TensorBase<1, T>(size),
+        m_data(size ? new T[size] : nullptr) {}
+    Tensor(const Shape<1>& shape) : Tensor(shape.dim[0]){}
+    ~Tensor() { delete[] m_data; }
+    Tensor& operator=(Tensor&& other) {
+        base_type::operator=(other);
+        m_data = other.m_data;
+    }
+
+    //! raw access into 1D array
+    T* data() { return m_data; }
+    const T* data() const { return m_data; }
+    //! raw indexing into 1D array
+    T& operator[](tensor_size_t i) { return m_data[i]; }
+    const T& operator[](tensor_size_t i) const { return m_data[i]; }
+
+    bool operator==(const Tensor<1, T>& other) const {
+        return base_type::operator==(other) &&
+            std::equal(m_data, m_data + base_type::size(), other.m_data);
+    }
+
+    bool operator!=(const Tensor<1, T>& other) const {
+        return !operator==(other); }
+
+    void swapWith(Tensor<1,T>& other)
+    {
+        TensorBase<1, T>::swapWith(other);
+        std::swap(m_data, other.m_data);
+    }
+
+    void swapDataWith(T*& new_data)
+    {
+        std::swap(m_data, new_data);
+    }
+
+    class TensorIterator
+    {
+        T* sub;
+    public:
+        // define all the stuff that std::lower_bound & co require...
+        using difference_type = std::size_t;
+        using value_type = T;
+        using pointer = T*;
+        using reference = T&;
+        using iterator_category = std::random_access_iterator_tag;
+
+        TensorIterator() = default;
+        TensorIterator(T* sub) : sub(sub) {}
+        TensorIterator& operator+=(tensor_size_t inc) { sub += inc; return *this; }
+        TensorIterator operator-(tensor_size_t inc) { return TensorIterator(sub - inc); }
+        TensorIterator& operator++() { ++sub; return *this; }
```

Comment:
I know a more recent revision of C++ added in some operator unification, but is another operator needed to ensure itr++ and ++itr are valid?